From 9b4ca998930c7b81b72e5068b3ebae33907ac4b2 Mon Sep 17 00:00:00 2001
From: Robert Griebl <rgriebl@trolltech.com>
Date: Wed, 27 Jan 2010 15:54:25 +0100
Subject: [PATCH] Always redraw the complete control when an input event comes in.

The problem here is that a pre-edit string won't be detected by
updateDisplayText(), so the control thinks nothing has changed when
a new pre-edit string is set.

Reviewed-By: Simon Hausmann
(cherry picked from commit 16f30906f6eea3e00351478555f153697a6e186d)
---
 src/gui/widgets/qlinecontrol.cpp |    6 +++---
 src/gui/widgets/qlinecontrol_p.h |    2 +-
 2 files changed, 4 insertions(+), 4 deletions(-)

--- a/src/gui/widgets/qlinecontrol.cpp
+++ b/src/gui/widgets/qlinecontrol.cpp
@@ -65,7 +65,7 @@ QT_BEGIN_NAMESPACE
     Updates the display text based of the current edit text
     If the text has changed will emit displayTextChanged()
 */
-void QLineControl::updateDisplayText()
+void QLineControl::updateDisplayText(bool forceUpdate)
 {
     QString orig = m_textLayout.text();
     QString str;
@@ -102,7 +102,7 @@ void QLineControl::updateDisplayText()
     m_textLayout.endLayout();
     m_ascent = qRound(l.ascent());
 
-    if (str != orig)
+    if (str != orig || forceUpdate)
         emit displayTextChanged(str);
 }
 
@@ -476,7 +476,7 @@ void QLineControl::processInputMethodEve
         }
     }
     m_textLayout.setAdditionalFormats(formats);
-    updateDisplayText();
+    updateDisplayText(/*force*/ true);
     if (cursorPositionChanged)
         emitCursorPositionChanged();
     if (isGettingInput)
--- a/src/gui/widgets/qlinecontrol_p.h
+++ b/src/gui/widgets/qlinecontrol_p.h
@@ -239,7 +239,7 @@ private:
     void init(const QString &txt);
     void removeSelectedText();
     void internalSetText(const QString &txt, int pos = -1, bool edited = true);
-    void updateDisplayText();
+    void updateDisplayText(bool forceUpdate = false);
 
     void internalInsert(const QString &s);
     void internalDelete(bool wasBackspace = false);
